### PR TITLE
Add policy to stop applications after a period has elapsed

### PR DESCRIPTION
**Pull request by sjcorbett**
Aka "the money saver".

`DurationSinceSensor` records the duration since it as added to an entity.

`StopAfterDurationPolicy` waits for an entity to reach a lifecycle, adds a duration-since sensor to it, waits for the duration to be greater than the configured lifetime, then stops the entity.

### CODE REVIEW

**Comment by sjcorbett** [comment]
Failure irrelevant: org.apache.brooklyn.AssemblyTest.checkBrooklynCoreFeature

**Comment by sjcorbett** [comment]
Example blueprint:
```yaml
location: localhost
services:
- type: org.apache.brooklyn.entity.stock.BasicApplication
  brooklyn.policies:
  - type: org.apache.brooklyn.policy.action.StopAfterDurationPolicy
    brooklyn.config:
      lifetime: 10s
      pollPeriod: 2s
  brooklyn.children:
  - type: org.apache.brooklyn.entity.software.base.EmptySoftwareProcess
```

File: core/src/main/java/org/apache/brooklyn/core/sensor/DurationSinceSensor.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        timeSupplier = params.get(TIME_SUPPLIER);
+        epochSensor = Sensors.newLongSensor(sensor.getName() + ".epoch");
+    }
+
+    @Override
+    public void apply(final EntityLocal entity) {
+        super.apply(entity);
+
+        if (entity.sensors().get(epochSensor) == null) {
+            entity.sensors().set(epochSensor, epochSupplier.get());
```
Comment (Graeme-Miller) [inline core/src/main/java/org/apache/brooklyn/core/sensor/DurationSinceSensor.java]: what if epochSupplier returns a null value here? Looks like null values are checked for in UpdateTimeSince , but is there any point creating function feed?

File: core/src/test/java/org/apache/brooklyn/core/sensor/DurationSinceSensorTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.brooklyn.core.test.entity.TestEntity;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.apache.brooklyn.util.time.Duration;
+import org.testng.annotations.Test;
+
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableMap;
+
+public class DurationSinceSensorTest extends BrooklynAppUnitTestSupport {
```
Comment (Graeme-Miller) [inline core/src/test/java/org/apache/brooklyn/core/sensor/DurationSinceSensorTest.java]: maybe add a test for when the epochSupplier/timeSupplier return null?

File: policy/src/main/java/org/apache/brooklyn/policy/action/StopAfterDurationPolicy.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            }
+        }
+    }
+
+    private class TimeIsUpListener implements SensorEventListener<Duration> {
+        @Override
+        public void onEvent(SensorEvent<Duration> event) {
+            synchronized (eventLock) {
+                if (!config().get(FIRED_STOP)) {
+                    if (!config().get(LIFETIME).subtract(event.getValue()).isPositive()) {
```
Comment (Graeme-Miller) [inline policy/src/main/java/org/apache/brooklyn/policy/action/StopAfterDurationPolicy.java]: It's a niggle, but is there an "isNegative" method and would this be clearer:
`config().get(LIFETIME).subtract(event.getValue()).isNegative()`

**Comment by Graeme-Miller** [comment]
couple of minor comments but LGTM

File: core/src/main/java/org/apache/brooklyn/core/sensor/DurationSinceSensor.java (hunk 1 above)
Comment (sjcorbett) [inline core/src/main/java/org/apache/brooklyn/core/sensor/DurationSinceSensor.java]: I think if epochSupplier returns null it should default to using CURRENT_TIME_SUPPLIER. It's surely a programming error. I've updated the description of the config key to indicate this.

**Comment by sjcorbett** [comment]
@Graeme-Miller Thanks for your comments.
